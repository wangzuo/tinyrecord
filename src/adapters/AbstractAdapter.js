import _ from 'lodash';
import * as Arel from 'arel';
import AlterTable from './schema/AlterTable';
import SchemaCreation from './schema/SchemaCreation';
import TableDefinition from './schema/TableDefinition';
import Table from './schema/Table';
import SchemaCache from './schema/SchemaCache';
import Relation from '../Relation';
import TypeMap from '../type/TypeMap';
import * as Type from '../Type';
import SqlTypeMetadata from './SqlTypeMetadata';
import QueryAttribute from '../relation/QueryAttribute';

class Version {}

class BindCollector extends Arel.collectors.Bind {
  compile(bvs, conn) {
    const castedBinds = bvs.map(x => x.valueForDatabase);
    return super.compile(castedBinds.map(value => conn.quote(value)));
  }
}

class SQLString extends Arel.collectors.SQLString {
  compile(bvs, conn) {
    return super.compile(bvs);
  }
}

export default class AbstractAdapter {
  static ADAPTER_NAME = 'Abstract';

  constructor(connection) {
    this.connection = connection;
    this.schemaCreation = new SchemaCreation(this);
    this.nativeDatabaseTypes = {};
    this.schemaCache = new SchemaCache(this);

    this.typeMap = new TypeMap();
    this.initializeTypeMap(this.typeMap);
    // todo
    this.preparedStatements = true;

    this.visitor = this.arelVisitor;
  }

  get arelVisitor() {
    return new Arel.visitors.ToSql(this);
  }

  get collector() {
    if (this.preparedStatements) {
      return new SQLString();
    }

    return new BindCollector();
  }

  async unpreparedStatement(block) {
    const oldPreparedStatements = this.preparedStatements;
    await block();

    return new Promise((resolve, reject) => {
      this.preparedStatements = false;

      block()
        .then(res => {
          this.preparedStatements = oldPreparedStatements;
          resolve(res);
        })
        .catch(err => {
          this.preparedStatements = oldPreparedStatements;
          reject(err);
        });
    });
  }

  initializeTypeMap(m) {
    this.registerClassWithLimit(m, /boolean/i, Type.Boolean);
    this.registerClassWithLimit(m, /char/i, Type.String);
    this.registerClassWithLimit(m, /binary/i, Type.Binary);
    this.registerClassWithLimit(m, /text/i, Type.Text);
    this.registerClassWithPrecision(m, /date/i, Type.Date);
    this.registerClassWithPrecision(m, /time/i, Type.Time);
    this.registerClassWithPrecision(m, /datetime/i, Type.DateTime);
    this.registerClassWithLimit(m, /float/i, Type.Float);
    this.registerClassWithLimit(m, /int/i, Type.Integer);

    m.aliasType(/blob/i, 'binary');
    m.aliasType(/clob/i, 'text');
    m.aliasType(/timestamp/i, 'datetime');
    m.aliasType(/numeric/i, 'decimal');
    m.aliasType(/number/i, 'decimal');
    m.aliasType(/double/i, 'float');

    m.registerType(/decimal/i, null, sqlType => {
      const scale = this.extractScale(sqlType);
      const precision = this.extractPrecision(sqlType);

      if (scale === 0) {
        return new Type.DecimalWithoutScale({ precision });
      } else {
        return new Type.Decimal({ precision, scale });
      }
    });

    return m;
  }

  registerClassWithLimit(mapping, key, klass) {
    mapping.registerType(key, null, (...args) => {
      const limit = this.extractLimit(_.last(args));
      return new klass({ limit });
    });
  }

  registerClassWithPrecision(mapping, key, klass) {
    mapping.registerType(key, null, (...args) => {
      const precision = this.extractPrecision(_.last(args));
      return new klass({ precision });
    });
  }

  extractPrecision(sqlType) {
    const m = sqlType.match(/\((\d+)(,\d+)?\)/);
    if (m) return _.toInteger(m[1]);
  }

  extractLimit(sqlType) {
    if (sqlType.match(/^bigint/i)) return 8;
    const m = sqlType.match(/\((.*)\)/);
    if (m) return m[1];
  }

  withoutPreparedStatement(binds) {
    return !this.preparedStatements || _.isEmpty(binds);
  }

  async toSql(arel, binds = []) {
    if (arel.ast) {
      // TODO resolve QueryAttribute type
      for (const bind of binds) {
        if (bind instanceof QueryAttribute) {
          bind.type = await bind.type;
        }
      }

      const collected = this.visitor.accept(arel.ast, this.collector);
      return collected.compile(binds, this);
    }

    return arel;
  }

  cacheableQuery(klass, arel) {
    const collected = this.visitor.accept(arel.ast, new SQLString());
    // if (this.preparedStatements) {
    //   return klass.query(collected.value);
    // }
    // return klass.partialQuery(collected.value);

    return klass.query(collected.value);
  }

  async select(sql, name = null, binds = []) {
    return await this.execQuery(sql, name, binds, { prepare: false });
  }

  async selectPrepared(sql, name = null, binds = []) {
    return await this.execQuery(sql, name, binds, { prepare: true });
  }

  async selectAll(arel, name = null, binds = [], options = {}) {
    let { preparable } = options || null;
    [arel, binds] = this.bindsFromRelation(arel, binds);
    const sql = await this.toSql(arel, binds);
    if (
      !this.preparedStatements ||
      (_.isString(arel) && _.isNull(preparable))
    ) {
      preparable = false;
    } else {
      preparable = this.visitor.preparable;
    }

    if (this.preparedStatements && preparable) {
      return await this.selectPrepared(sql, name, binds);
    }

    return await this.select(sql, name, binds);
  }

  bindsFromRelation(relation, binds) {
    if (relation instanceof Relation && _.isEmpty(binds)) {
      relation = relation.arel;
      binds = relation.boundAttributes;
    }
    return [relation, binds];
  }

  async selectOne(arel, name = null, binds = []) {}

  async selectValue(arel, name = null, binds = []) {}

  async selectValues(arel, name = null, binds = []) {
    const result = await this.selectRows(arel, name, binds);
    return result.rows.map(row => row[0]);
  }

  async selectRows(arel, name = null, binds = []) {
    return this.selectAll(arel, name, binds);
  }

  async execute(sql, name = null) {}

  async insert(
    arel,
    name = null,
    pk = null,
    idValue = null,
    sequenceName = null,
    binds = []
  ) {
    const value = await this.execInsert(
      await this.toSql(arel, binds),
      name,
      binds,
      pk,
      sequenceName
    );

    return idValue || this.lastInsertedId(value);
  }

  sqlForInsert(sql, pk, idValue, sequenceName, binds) {
    return [sql, binds];
  }

  async update(arel, name = null, binds = []) {
    const sql = await this.toSql(arel, binds);
    return await this.execUpdate(sql, name, binds);
  }

  async delete(arel, name = null, binds = []) {
    return this.execDelete(await this.toSql(arel, binds), name, binds);
  }

  async execInsert(
    sql,
    name = null,
    binds = [],
    pk = null,
    sequenceName = null
  ) {
    const result = this.sqlForInsert(sql, pk, null, sequenceName, binds);
    return this.execQuery(result[0], name, result[1]);
  }

  async execDelete(sql, name = null, binds = []) {
    return this.execQuery(sql, name, binds);
  }

  async truncate(tableName, name = null) {}

  async execUpdate(sql, name = null, binds = []) {
    return await this.execQuery(sql, name, binds);
  }

  async execQuery(sql, name = 'SQL', binds = []) {
    throw new Error('Not implemented error');
  }

  dataSourceSql(name = null, options = {}) {
    throw new Error('Not implemented error');
  }

  quotedScope(name = null, options = {}) {
    throw new Error('Not implemented error');
  }

  async dataSources() {
    try {
      return await this.selectValues(this.dataSourceSql(), 'SCHEMA');
    } catch (e) {
      throw e;
    }
  }

  async dataSourceExists() {
    try {
      return await this.selectValues(this.dataSourceSql(name), 'SCHEMA');
    } catch (e) {
      throw e;
    }
  }

  async tables() {
    return await this.selectValues(
      this.dataSourceSql(null, { type: 'BASE TABLE' }),
      'SCHEMA'
    );
  }

  async tableExists(tableName) {
    const tables = await this.selectValues(
      this.dataSourceSql(tableName, { type: 'BASE TABLE' }),
      'SCHEMA'
    );
    return !_.isEmpty(tables);
  }

  async views() {
    return await this.selectValues(
      this.dataSourceSql(null, { type: 'VIEW' }),
      'SCHEMA'
    );
  }
  async viewExists(viewName) {
    await this.selectValues(
      this.dataSourceSql(viewName, { type: 'VIEW' }),
      'SCHEMA'
    );
  }

  async indexes() {
    throw new Error('#indexes is not implemented');
  }

  async indexExists(tableName, columnName, options = {}) {}

  async columns(tableName) {
    const columnDefinitions = await this.columnDefinitions(tableName);

    return columnDefinitions.map(field =>
      this.newColumnFromField(tableName, field)
    );
  }

  async columnExists(tableName, columnName, type = null, options = {}) {}

  primaryKey(tableName) {
    const pk = this.primaryKeys(tableName);
  }

  async createTable(tableName, options = {}, block) {
    const comment = options.comment || null;
    const td = this.createTableDefinition(
      tableName,
      options.temporary,
      options.options,
      options.as,
      { comment }
    );

    if (options.id !== false && !options.as) {
      const pk = options.primaryKey || 'id'; // todo: getPrimaryKey

      if (_.isArray(pk)) {
        td.primaryKeys(pk);
      } else {
        td.primaryKey(pk, options.id || options.primaryKey, options);
      }
    }

    if (block) block(td);

    if (options.force) {
      await this.dropTable(tableName, { ...options, ifExists: true });
    }

    const sql = this.schemaCreation.accept(td);
    await this.execute(sql);
  }

  async changeTable(tableName, options = {}) {}

  async renameTable(tableName, newName) {
    throw new Error('rename_table is not implemented');
  }

  async dropTable(tableName, options = {}) {
    const sql = `DROP TABLE${options.ifExists
      ? ' IF EXISTS'
      : ''} ${this.quoteTableName(tableName)}`;
    await this.execute(sql);
  }

  async addColumn(tableName, columnName, type, options = {}) {
    const at = this.createAlterTable(tableName);
    at.addColumn(columnName, type, options);
    const sql = this.schemaCreation.accept(at);
    await this.execute(sql);
  }

  async removeColumns(tableName, ...columnNames) {
    if (_.isEmpty(columnNames)) {
      throw new Error(
        `You must specify at least one column name. Example: remove_columns(:people, :first_name)`
      );
    }

    for (const columnName of columnNames) {
      await this.removeColumn(tableName, columnName);
    }
  }

  async removeColumn(tableName, columnName, type = null, options = {}) {
    const sql = `ALTER TABLE ${this.quoteTableName(
      tableName
    )} DROP ${this.quoteColumnName(columnName)}`;

    await this.execute(sql);
  }

  changeColumn(tableName, columnName, type, options = {}) {
    throw new Error('change_column is not implemented');
  }

  changeColumnDefault(tableName, columnName, defaultOrChanges) {}

  changeColumnNull(tableName, columnName, nil, _default = null) {
    throw new Error('changeColumnNull is not implemented');
  }

  renameColumn(tableName, columnName, newColumnName) {
    throw new Error('renameColumn is not implemented');
  }

  addIndex(tableName, columnName, options = {}) {}
  removeIndex(tableName, options = {}) {}
  renameIndex(tableName, oldName, newName) {}
  indexName(tableName, options) {}
  indexNameExists(tableName, indexName, _default = null) {}

  addReference(tableName, refName, options) {}
  removeReference(tableName, refName) {}

  async addTimestamps(tableName, options = {}) {
    options.null = options.null || false;

    await this.addColumn(tableName, 'created_at', 'datetime', options);
    await this.addColumn(tableName, 'updated_at', 'datetime', options);
  }

  async removeTimestamps(tableName, options = {}) {
    await this.removeColumn(tableName, 'updated_at');
    await this.removeColumn(tableName, 'created_at');
  }

  foreignKeys(tableName) {
    throw new Error('foreignKeys is not implemented');
  }

  addForeignKey(fromTable, toTable, options = {}) {}
  removeForeignKey(fromTable) {}
  foreignKeyExists(fromTable) {}
  foreignKeyColumnFor(tableName) {}
  foreignKeyOptions(fromTable, toTable) {}

  dumpSchemaInformatin() {}
  initializeSchemaMigrationsTable() {}
  initializeInternalMetadataTable() {}

  internalStringOptionsForPrimaryKey() {
    return { primaryKey: true };
  }
  assume_migrated_upto_version() {}

  // quoting

  quote(value) {
    return this._quote(value);
  }

  quoteString(s) {
    return s;
  }

  _quote(value) {
    if (_.isString(value)) {
      return `'${this.quoteString(value)}'`;
    } else if (value === true) {
      return this.quotedTrue;
    } else if (value == false) {
      return this.quotedFalse;
    } else if (_.isNull(value)) {
      return 'NULL';
    }

    return value;
    // return `'${value}'`;
  }

  typeCastedBinds(binds) {
    if (_.isArray(binds[0])) {
      return binds.map(([column, value]) => this.typeCast(value, column));
    }
    return binds.map(attr => this.typeCast(attr.valueForDatabase));
  }

  typeCast(value, column = null) {
    // const value = idValueForDatabase(value);

    return this._typeCast(value);
  }

  _typeCast(value) {
    if (_.isString(value)) {
      return value;
    } else if (value === true) {
      return this.unquotedTrue;
    } else if (value === false) {
      return this.unquotedFalse;
    }

    return value;
  }

  quoteTableName(tableName) {
    return `"${tableName}"`;
  }

  quoteColumnName(columnName) {
    return `"${columnName}"`;
  }

  quoteDefaultExpression() {}

  createTableDefinition(...args) {
    return new TableDefinition(...args);
  }

  createAlterTable(name) {
    return new AlterTable(this.createTableDefinition(name));
  }

  updateTableDefinition(tableName, base) {
    return new Table(tableName, base);
  }

  typeToSql(type, options = {}) {
    const native = this.nativeDatabaseTypes[type];

    if (native) {
      let columnTypeSql = native.name || native;
      const limit = options.limit || native.limit;
      const precision = options.precision || native.precision;
      const scale = options.scale || native.scale;

      if (type === 'decimal') {
        if (precision) {
          columnTypeSql += scale ? `(${precision},${scale})` : `"${precision}"`;
        } else if (scale) {
          throw new Error(
            'Error adding decimal column: precision cannot be empty if scale is specified'
          );
        }
      } else if (
        _.includes(['datetime', 'timestamp', 'time', 'interval'], type) &&
        precision
      ) {
        if (precision >= 0 && precision <= 6) {
          columnTypeSql += `(${precision})`;
        } else {
          throw new Error(
            `No ${native.name} type has precision of ${precision}. The allowed range of precision is from 0 to 6`
          );
        }
      } else if (type !== 'primaryKey' && limit) {
        columnTypeSql += `(${limit})`;
      }

      return columnTypeSql;
    }

    return type;
  }

  fetchTypeMetadata(sqlType) {
    const castType = this.lookupCastType(sqlType);

    return new SqlTypeMetadata({
      sqlType,
      type: castType.type,
      limit: castType.limit,
      precision: castType.precision,
      scale: castType.scale
    });
  }

  lookupCastTypeFromColumn(column) {
    return this.lookupCastType(column.sqlType);
  }

  lookupCastType(sqlType) {
    return this.typeMap.lookup(sqlType);
  }

  combineBindParameters(options = {}) {
    const {
      fromClause,
      joinClause,
      whereClause,
      havingClause,
      limit,
      offset
    } = Object.assign(
      {
        fromClause: [],
        joinClause: [],
        whereClause: [],
        havingClause: [],
        limit: null,
        offset: null
      },
      options
    );

    const result = [
      ...fromClause,
      ...joinClause,
      ...whereClause,
      ...havingClause
    ];

    if (limit) result.push(limit);
    if (offset) result.push(offset);

    return result;
  }

  sanitizeLimit(limit) {
    // todo
    return _.toNumber(limit);
  }
}
