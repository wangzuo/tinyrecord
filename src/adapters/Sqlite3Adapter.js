import _ from 'lodash';
import AbstractAdapter from './AbstractAdapter';
import Result from '../Result';
import Column from './Column';

const NATIVE_DATABASE_TYPES = {
  primaryKey: 'INTEGER PRIMARY KEY AUTOINCREMENT NOT NULL',
  string: { name: 'varchar' },
  text: { name: 'text' },
  integer: { name: 'integer' },
  float: { name: 'float' },
  decimal: { name: 'decimal' },
  datetime: { name: 'datetime' }, // todo: datetime not null
  time: { name: 'time' },
  date: { name: 'date' },
  binary: { name: 'blob' },
  boolean: { name: 'boolean' }
};

export default class Sqlite3Adapter extends AbstractAdapter {
  static ADAPTER_NAME = 'SQLite';

  constructor(connection, logger, connectionOptions, config) {
    super(connection, logger, config);
    const sqlite3 = require('sqlite3');

    this.connection = new sqlite3.Database(':memory:'); // todo
    this.nativeDatabaseTypes = NATIVE_DATABASE_TYPES;
    this.supportsDdlTransactions = true;
    this.supportSavepoints = true;

    // alias
    this.columnDefinitions = this.tableStructure;
    this.addBelongsTo = this.addReference;
  }

  async explain(arel, binds = []) {
    sql = `EXPLAIN QUERY PLAN ${await this.toSql(arel, binds)}`;
    return new ExplainPrettyPrinter().pp(
      await this.execQuery(sql, 'EXPLAIN', [])
    );
  }

  // todo: try {} catch {}
  async execQuery(sql, name = null, binds = [], options = {}) {
    const prepare = options.prepare || false;
    const typeCastedBinds = this.typeCastedBinds(binds);

    console.log('execQuery', sql, typeCastedBinds);

    if (!prepare) {
      const stmt = this.connection.prepare(sql);

      if (!this.withoutPreparedStatement(binds)) {
        stmt.bind(typeCastedBinds);
      }

      return new Promise((resolve, reject) => {
        stmt.all((err, rows) => {
          if (err) return reject(err);

          // todo: empty when result empty
          const cols = rows.map(row => _.keys(row))[0];
          const records = rows.map(row => _.values(row));

          return resolve(new Result(cols, records));
        });

        stmt.finalize();
      });
    } else {
    }
  }

  async execInsert(
    sql,
    name = null,
    binds = [],
    pk = null,
    sequenceName = null
  ) {
    [sql, binds] = this.sqlForInsert(sql, pk, null, sequenceName, binds);

    const stmt = this.connection.prepare(sql);
    const typeCastedBinds = this.typeCastedBinds(binds);
    stmt.bind(typeCastedBinds);

    return new Promise((resolve, reject) => {
      stmt.run(function(err) {
        if (err) return reject(err); // todo: error

        const result = new Result();
        result.lastInsertRowId = this.lastID;

        return resolve(result);
      });
    });
  }

  async execDelete(sql, name = 'SQL', binds = []) {
    return await this.execQuery(sql, name, binds);
  }

  async execUpdate(sql, name = 'SQL', binds = []) {
    console.log(sql, binds);
    // return await this.execDelete(...args);
  }

  lastInsertedId(result) {
    return result.lastInsertRowId;
  }

  async execute(sql, name = null) {
    console.log('execute', sql);

    return new Promise((resolve, reject) => {
      this.connection.run(sql, (err, response) => {
        if (err) return reject(err);
        return resolve(response);
      });
    });
  }

  beginDbTransaction() {
    this.log('begin transaction', null, () => {
      this.connection.transaction();
    });
  }

  commitDbTransaction() {
    this.log('commit transaction', null, () => {
      this.connection.commit();
    });
  }

  execRollbackDbTransaction() {
    this.log('rollback transaction', null, () => {
      this.connection.rollback();
    });
  }

  primaryKeys(tableName) {
    const pks = this.tableStructure(tableName).filter(f => f.pk > 0);
    return pks.sort((a, b) => a.pk - b.pk).map(f => f.name);
  }

  async removeIndex(tableName, options = {}) {
    const indexName = this.indexNameForRemove(tableName, options);
    await this.execQuery(`DROP INDEX ${this.quoteColumn_Name(indexName)}`);
  }

  async renameTable(tableName, newName) {
    await this.execQuery(
      `ALTER TABLE ${this.quoteTableName(
        tableName
      )} RENAME TO ${this.quoteTableName(newName)}`
    );
    return await this.renameTableIndexes(tableName, newName);
  }

  newColumnFromField(tableName, field) {
    let _default = null; // todo

    const typeMetadata = this.fetchTypeMetadata(field.type);
    return new Column(
      field.name,
      _default,
      typeMetadata,
      field.notnull === 0,
      tableName,
      null,
      field.collation
    );
  }

  dataSourceSql(name = null, options = {}) {
    const type = options.type || null;
    const scope = this.quotedScope(name, { type });
    scope.type = scope.type || "'table','view'";

    let sql = "SELECT name FROM sqlite_master WHERE name <> 'sqlite_sequence'";
    if (scope.name) {
      sql += ` AND name = ${scope.name}`;
    }
    sql += ` AND type IN (${scope.type})`;
    return sql;
  }

  quotedScope(name, options = {}) {
    let type = options.type || null;

    if (type === 'BASE TABLE') type = "'table'";
    else if (type === 'VIEW') type = "'view'";
    const scope = {};
    if (name) scope.name = this.quote(name);
    if (type) scope.type = type;

    return scope;
  }

  async tableStructure(tableName) {
    const structure = await this.execQuery(
      `PRAGMA table_info(${this.quoteTableName(tableName)})`,
      'SCHEMA'
    );

    if (_.isEmpty(structure)) {
      throw new Error(`Could not find table ${tableName}`);
    }

    return await this.tableStructureWithCollation(tableName, structure);
  }

  async tableStructureWithCollation(tableName, basicStructure) {
    const collationHash = {};
    const sql = `SELECT sql FROM (SELECT * FROM sqlite_master UNION ALL SELECT * FROM sqlite_temp_master) WHERE type = 'table' AND name = ${this.quote(
      tableName
    )}`;

    const result = (await this.execQuery(sql, 'SCHEMA')).first();

    if (result) {
      // todo
      const columnString = _.last(result.sql.split('('));
      columnString.split(',').forEach(columnString => {
        const m = columnString.match(/.*\"(\w+)\".*collate\s+\"(\w+)\".*/);

        if (m) {
          collationHash[m[1]] = m[2];
        }
      });

      // todo
      // return basicStructure.map(column => {
      //   const columnName = column.name;

      //   return column;
      // });

      return basicStructure.hashRows;
    }
  }
}
