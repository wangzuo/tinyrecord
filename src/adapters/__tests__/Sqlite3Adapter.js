import Base from '../../Base';
import { testAdapter } from './AbstractAdapter';
import Sqlite3Adapter from '../Sqlite3Adapter';

class User extends Base {
  static tableName = 'users';
}

beforeAll(() => {
  Base.establishConnection({
    adapter: 'sqlite3',
    database: ':memory:'
  });

  return Base.connection.createTable('users', { force: true }, t => {
    t.string('name', { default: 'untitled' });
    t.string('email');
    t.integer('age', { default: 0 });
    t.timestamps();
  });
});

test('ADAPTER_NAME', () => {
  expect(Sqlite3Adapter.ADAPTER_NAME).toBe('SQLite');
});

test('logger', () => {
  expect(User.connection.logger).not.toBeNull();
});

test('loadSchema', async () => {
  await User.loadSchema();
});

// todo: test views
test('views', async () => {
  const views = await Base.connection.views();
  expect(views).toEqual([]);
});

test('tableExists', async () => {
  expect(await Base.connection.tableExists('users')).toBe(true);
  expect(await Base.connection.tableExists('people')).toBe(false);
});

test('attributeNames', async () => {
  expect(await User.attributeNames()).toEqual([
    'id',
    'name',
    'email',
    'age',
    'created_at',
    'updated_at'
  ]);
});

test('attributeTypes', async () => {
  expect(await User.attributeTypes()).toMatchSnapshot();
});

test('lookupCastType', () => {
  expect(User.connection.lookupCastType('INTEGER')).toMatchSnapshot();
});

test('tableStructure', async () => {
  expect(await User.connection.tableStructure('users')).toMatchSnapshot();
});

test('records', async () => {
  expect(await User.all.records()).toMatchSnapshot();
  await User.create({ name: 'test', email: 'test@example.com' });
  expect(await User.all.records()).toMatchSnapshot();
});

// test('having', async () => {});

testAdapter(Base);
