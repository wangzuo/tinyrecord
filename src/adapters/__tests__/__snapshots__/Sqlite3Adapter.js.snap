// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributeTypes 1`] = `
Object {
  "age": Integer {
    "limit": null,
    "precision": null,
    "scale": null,
  },
  "created_at": DateTime {
    "limit": null,
    "precision": null,
    "scale": null,
  },
  "email": String {
    "limit": null,
    "precision": null,
    "scale": null,
  },
  "id": Integer {
    "limit": null,
    "precision": null,
    "scale": null,
  },
  "name": String {
    "limit": null,
    "precision": null,
    "scale": null,
  },
  "updated_at": DateTime {
    "limit": null,
    "precision": null,
    "scale": null,
  },
}
`;

exports[`columns 1`] = `
Array [
  Column {
    "collation": null,
    "comment": null,
    "default": null,
    "defaultFunction": null,
    "name": "id",
    "nil": false,
    "sqlTypeMetaData": SqlTypeMetadata {
      "limit": null,
      "precision": null,
      "scale": null,
      "sqlType": "INTEGER",
      "type": "integer",
    },
    "tableName": "users",
  },
  Column {
    "collation": null,
    "comment": null,
    "default": "untitled",
    "defaultFunction": null,
    "name": "name",
    "nil": true,
    "sqlTypeMetaData": SqlTypeMetadata {
      "limit": null,
      "precision": null,
      "scale": null,
      "sqlType": "varchar",
      "type": null,
    },
    "tableName": "users",
  },
  Column {
    "collation": null,
    "comment": null,
    "default": null,
    "defaultFunction": null,
    "name": "email",
    "nil": true,
    "sqlTypeMetaData": SqlTypeMetadata {
      "limit": null,
      "precision": null,
      "scale": null,
      "sqlType": "varchar",
      "type": null,
    },
    "tableName": "users",
  },
  Column {
    "collation": null,
    "comment": null,
    "default": "0",
    "defaultFunction": null,
    "name": "age",
    "nil": true,
    "sqlTypeMetaData": SqlTypeMetadata {
      "limit": null,
      "precision": null,
      "scale": null,
      "sqlType": "integer",
      "type": "integer",
    },
    "tableName": "users",
  },
  Column {
    "collation": null,
    "comment": null,
    "default": null,
    "defaultFunction": null,
    "name": "created_at",
    "nil": false,
    "sqlTypeMetaData": SqlTypeMetadata {
      "limit": null,
      "precision": null,
      "scale": null,
      "sqlType": "datetime",
      "type": null,
    },
    "tableName": "users",
  },
  Column {
    "collation": null,
    "comment": null,
    "default": null,
    "defaultFunction": null,
    "name": "updated_at",
    "nil": false,
    "sqlTypeMetaData": SqlTypeMetadata {
      "limit": null,
      "precision": null,
      "scale": null,
      "sqlType": "datetime",
      "type": null,
    },
    "tableName": "users",
  },
]
`;

exports[`group 1`] = `"SELECT \\"users\\".* FROM \\"users\\" GROUP BY name"`;

exports[`joins string 1`] = `"SELECT \\"users\\".* FROM \\"users\\" LEFT JOIN bookmarks ON bookmarks.user_id = users.id"`;

exports[`limit 1`] = `"SELECT  \\"users\\".* FROM \\"users\\" LIMIT 1"`;

exports[`limit 2`] = `"SELECT  \\"users\\".* FROM \\"users\\" WHERE \\"users\\".\\"name\\" = 'test' LIMIT 1"`;

exports[`lookupCastType 1`] = `
Integer {
  "limit": null,
  "precision": null,
  "scale": null,
}
`;

exports[`new 1`] = `
Object {
  "age": FromDatabase {
    "_value": 0,
    "name": "age",
    "originalAttribute": Object {},
    "type": Integer {
      "limit": null,
      "precision": null,
      "scale": null,
    },
    "valueBeforeTypeCast": "0",
  },
  "created_at": FromDatabase {
    "name": "created_at",
    "originalAttribute": Object {},
    "type": DateTime {
      "limit": null,
      "precision": null,
      "scale": null,
    },
    "valueBeforeTypeCast": null,
  },
  "email": FromUser {
    "_value": "new@example.com",
    "name": "email",
    "originalAttribute": Object {},
    "type": String {
      "limit": null,
      "precision": null,
      "scale": null,
    },
    "valueBeforeTypeCast": "new@example.com",
  },
  "id": FromDatabase {
    "name": "id",
    "originalAttribute": Object {},
    "type": Integer {
      "limit": null,
      "precision": null,
      "scale": null,
    },
    "valueBeforeTypeCast": null,
  },
  "name": FromUser {
    "_value": "new",
    "name": "name",
    "originalAttribute": Object {},
    "type": String {
      "limit": null,
      "precision": null,
      "scale": null,
    },
    "valueBeforeTypeCast": "new",
  },
  "updated_at": FromDatabase {
    "name": "updated_at",
    "originalAttribute": Object {},
    "type": DateTime {
      "limit": null,
      "precision": null,
      "scale": null,
    },
    "valueBeforeTypeCast": null,
  },
}
`;

exports[`new 2`] = `
Object {
  "age": 0,
  "created_at": undefined,
  "email": "new@example.com",
  "id": undefined,
  "name": "new",
  "updated_at": undefined,
}
`;

exports[`offset 1`] = `"SELECT \\"users\\".* FROM \\"users\\" ORDER BY name ASC LIMIT -1 OFFSET 10"`;

exports[`order 1`] = `"SELECT \\"users\\".* FROM \\"users\\" ORDER BY \\"users\\".\\"name\\" ASC"`;

exports[`order 2`] = `"SELECT \\"users\\".* FROM \\"users\\" ORDER BY \\"users\\".\\"email\\" DESC"`;

exports[`order 3`] = `"SELECT \\"users\\".* FROM \\"users\\" ORDER BY \\"users\\".\\"name\\" ASC, \\"users\\".\\"email\\" DESC"`;

exports[`order 4`] = `"SELECT \\"users\\".* FROM \\"users\\" ORDER BY name"`;

exports[`order 5`] = `"SELECT \\"users\\".* FROM \\"users\\" ORDER BY name DESC"`;

exports[`order 6`] = `"SELECT \\"users\\".* FROM \\"users\\" ORDER BY name DESC, email"`;

exports[`records 1`] = `Array []`;

exports[`records 2`] = `
Array [
  User {
    "_newRecord": false,
    "_startTransactionState": Object {},
    "attributes": Object {},
    "destroyed": false,
    "destroyedByAssociation": null,
    "markedForDestruction": false,
    "proxy": [Circular],
    "readonly": false,
    "transactionState": null,
  },
]
`;

exports[`select 1`] = `"SELECT \\"users\\".\\"name\\", \\"users\\".\\"email\\", email as user_email FROM \\"users\\""`;

exports[`table.type 1`] = `
String {
  "limit": null,
  "precision": null,
  "scale": null,
}
`;

exports[`table.type 2`] = `
String {
  "limit": null,
  "precision": null,
  "scale": null,
}
`;

exports[`tableStructure 1`] = `
Array [
  Object {
    "cid": 0,
    "dflt_value": null,
    "name": "id",
    "notnull": 1,
    "pk": 1,
    "type": "INTEGER",
  },
  Object {
    "cid": 1,
    "dflt_value": "'untitled'",
    "name": "name",
    "notnull": 0,
    "pk": 0,
    "type": "varchar",
  },
  Object {
    "cid": 2,
    "dflt_value": null,
    "name": "email",
    "notnull": 0,
    "pk": 0,
    "type": "varchar",
  },
  Object {
    "cid": 3,
    "dflt_value": "0",
    "name": "age",
    "notnull": 0,
    "pk": 0,
    "type": "integer",
  },
  Object {
    "cid": 4,
    "dflt_value": null,
    "name": "created_at",
    "notnull": 1,
    "pk": 0,
    "type": "datetime",
  },
  Object {
    "cid": 5,
    "dflt_value": null,
    "name": "updated_at",
    "notnull": 1,
    "pk": 0,
    "type": "datetime",
  },
]
`;

exports[`tables 1`] = `
Array [
  "users",
  "posts",
]
`;

exports[`where 1`] = `"SELECT \\"users\\".* FROM \\"users\\""`;

exports[`where 2`] = `"SELECT \\"users\\".* FROM \\"users\\" WHERE \\"users\\".\\"name\\" = 'test'"`;

exports[`where 3`] = `"SELECT \\"users\\".* FROM \\"users\\" WHERE \\"users\\".\\"name\\" = 'test' AND \\"users\\".\\"email\\" = 'test@example.com'"`;

exports[`where 4`] = `"SELECT \\"users\\".* FROM \\"users\\" WHERE \\"users\\".\\"name\\" = 'test' AND \\"users\\".\\"email\\" = 'test@example.com'"`;

exports[`where 5`] = `"SELECT \\"users\\".* FROM \\"users\\" WHERE (name = 'test') AND (age = 10)"`;

exports[`where 6`] = `"SELECT \\"users\\".* FROM \\"users\\" WHERE (id = 1)"`;
